### PR TITLE
[JENKINS-60092] - Prevent NullPointerException in ViewGroupMixIn#getPrimaryView() if no view is defined, return null instead

### PR DESCRIPTION
**Pull request by pilou-**
See [JENKINS-60092](https://issues.jenkins-ci.org/browse/JENKINS-60092).

This exception occurs while defining a view using job DSL:

    SEVERE  jenkins.InitReactorRunner$1#onTaskFailed: Failed ConfigurationAsCode.init
    java.lang.ArrayIndexOutOfBoundsException: 0
        at java.util.concurrent.CopyOnWriteArrayList.get(CopyOnWriteArrayList.java:388)
        at java.util.concurrent.CopyOnWriteArrayList.get(CopyOnWriteArrayList.java:397)
        at hudson.model.ViewGroupMixIn.getPrimaryView(ViewGroupMixIn.java:172)
        at hudson.model.ViewGroupMixIn.getView(ViewGroupMixIn.java:134)
        at jenkins.model.Jenkins.getView(Jenkins.java:1834)
        at javaposse.jobdsl.plugin.JenkinsJobManagement.createOrUpdateView(JenkinsJobManagement.java:175)
        at javaposse.jobdsl.dsl.JobManagement$createOrUpdateView$2.call(Unknown Source)
        at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:141)
        at javaposse.jobdsl.dsl.AbstractDslScriptLoader$_extractGeneratedViews_closure5.doCall(AbstractDslScriptLoader.groovy:216)

I didn't find where to add a unit test (it should be executed before `Jenkins.InitReactorRunner$1#onAttained: Completed initialization`).

### Proposed changelog entries

* Entry 1: Prevent NullPointerException in ViewGroupMixIn#getPrimaryView() if no view is defined, return null instead

Original:  JENKINS-60092, avoid exception when views are defined using job DSL only

### Submitter checklist

- [X] JIRA issue is well described
- [X] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### CODE REVIEW

**Review by timja — APPROVED** [review]

**Comment by oleg-nenashev** [comment]
For reviewers: I tried to cleanup nullness of primary views in few PRs before, e.g. in https://github.com/jenkinsci/jenkins/pull/3156 and https://github.com/jenkinsci/jenkins/pull/2815 . There is still some mess with them in APIs, and usages generally do not expect `getPrimaryView()` to return null

Example: https://github.com/jenkinsci/jenkins/blob/41a13dffc612ca3b5c48ab3710500562a3b40bf7/core/src/main/java/jenkins/model/Jenkins.java#L2269

Personally I think we should bite the bullet and say that the API is nullable, because there are already plugins which return null in relevant methods

**Review by res0nance — CHANGES_REQUESTED** [review]
Since it can return null can you annotate the function with `@CheckForNull` so that users of the api do the appropriate checks

File: core/src/main/java/hudson/model/ViewGroupMixIn.java [hunk 1]
```diff
@@ -163,12 +163,14 @@ public View getView(@CheckForNull String name) {
     }
 
     /**
-     * Returns the primary {@link View} that renders the top-page of Hudson.
+     * Returns the primary {@link View} that renders the top-page of Hudson or
+     * {@code null} if there is no primary one defined.
      */
     @Exported
+    @CheckForNull
```
Comment (timja) [inline core/src/main/java/hudson/model/ViewGroupMixIn.java]: is there anywhere in jenkins core that needs updating to now handle the nullness?

**Comment by timja** [comment]
Please don't force push while under review, it makes diffs harder to understand and github notifications don't handle it well at all it just says something changed.

**Comment by pilou-** [comment]
PR updated: `CheckForNull` added.

Is the better way to test this change to create a dedicated plugin which would check that null is returned (and use it `@WithPlugin`) ?

**Comment by timja** [comment]
> PR updated: `CheckForNull` added.
> 
> Is the better way to test this change to create a dedicated plugin which would check that null is returned (and use it `@WithPlugin`) ?

I would search for usages in core first and see if anywhere need updating.

Spotbugs should catch it but afaik spotbugs isn't currently failing the build here

File: core/src/main/java/hudson/model/ViewGroupMixIn.java (hunk 1 above)
Comment (jvz) [reply core/src/main/java/hudson/model/ViewGroupMixIn.java]: Wouldn't that cause a compile error from spotbugs now that the annotation is there?

**Review by varyvol — APPROVED** [review]
No Spotbugs error and there's not real behaviour change related to nullability so looks good to me.

**Review by alecharp — APPROVED** [review]
I didn't see any errors or warnings from spotbugs analysis, so :+1:

**Comment by varyvol** [comment]
Closing to retrigger CI build.

**Review by oleg-nenashev — COMMENTED** [review]
It is a "pick your poison" situation. Whatever we do, there is code in downstream which needs fixes. Generally we need an architectural decision for "Can we have a view group without views?". I would rather vouch for saying "no", but some plugins de-facto do that.

Maybe the best solution would be to create a `NoViewsView` object to raise the issue without returning null, but it is also quite a hack. Feedback from other @jenkinsci/core members will be appreciated

**Comment by daniel-beck** [comment]
> some plugins de-facto do that

Which plugins? If we always consider all 1600+ plugins we're always going with the worst common denominator.

**Review by jglick — APPROVED** [review]
Guess this cannot make things worse, but really `makeSearchIndex` and `doSubmitDescription` (at least) should be fixed to match. Generally it is a configuration error to have no views; you should always have at least `AllView`. Perhaps we should enforce this invariant earlier.

**Comment by daniel-beck** [comment]
@jglick The ability to remove a non-default `AllView` is a feature IMO.

**Comment by jglick** [comment]
> The ability to remove a non-default `AllView` is a feature IMO.

Yes. I should have rephrased: perhaps we should enforce an invariant that there is always at least one view, with `AllView` used as the final fallback.

**Comment by oleg-nenashev** [comment]
I agree with using `AllView` as a fallback.
IIRC I tried it also and hit some other obstacles, but I do not remember which ones.
Anyway, it can be done in a follow-up.

**Comment by oleg-nenashev** [comment]
I plan to merge it tomorrow if no negative feedback

>> some plugins de-facto do that

> Which plugins? If we always consider all 1600+ plugins we're always going with the worst common denominator.

Agreed.

**Comment by oleg-nenashev** [comment]
Thanks @pilou- , it is much appreciated ! 🙇

**Comment by oleg-nenashev** [comment]
Created https://issues.jenkins-ci.org/browse/JENKINS-60247 as a follow-up improvement

**Pull request merged**